### ISSUE TITLE
[Tracking] moe-store split: param-server/ckpt store repo + pure offloading engine

### ISSUE THREAD

**Issue by drunkcoding** [EfficientMoE/MoE-Infinity#223]
## Prerequisites

- [x] I have searched existing issues and reviewed documentation.

## Problem Description

Per the decisions in #222, the upcoming **Omni-Infinity** project will share
MoE-Infinity's weight-store infrastructure through a new library
**`EfficientMoE/moe-store`** rather than copy-then-diverge.

This repo already contains a **locked, evidence-verified execution plan** for
exactly this split: [`.sisyphus/plans/moe-store-split.md`]
(baseline `main@6da092f`, file:line-verified current-state evidence,
user-confirmed decision table). That plan's §5 requires a tracking issue to be
opened before PR1 lands. **This is that tracking issue.** This issue does not
redefine scope; it records the plan's scope and its per-PR verification gates
for reviewers.

> Correction note: the first version of this issue proposed a different
> boundary (moving `core/memory`/`core/prefetch`/transfer into the shared
> library). That contradicted the locked plan and is withdrawn. The
> authoritative boundary is below.

## Scope (locked, from `moe-store-split.md` §0)

**`EfficientMoE/moe-store` (new repo) owns model-architecture knowledge and
the on-disk store:**

- Checkpoint conversion (state-dict walk, v5 batched-expert remap, tensor-id
  assignment, topology derivation from `moe_infinity/runtime/model_offload.py`)
- Model registry (`moe_infinity/common/constants.py`), expert-id/dtype parsing
  (`moe_infinity/utils/hf_config.py`), `moe_infinity/utils/checkpoints.py`
- Sync\* model wrappers (`moe_infinity/models/*`, incl. `deepseek_v4/`)
- The `core/aio/` C++ tensor store, with a **new arch-aware v2 format**:
  one group per topology node (expert node = one expert's matrices in slot
  order), invariants G1–G4 (no partition straddle; 4KB group alignment;
  registry slot order, never sorted; quant weight+scales co-grouped) → every
  expert fetch is **one read at `(offset, size)`**
- CLI: `moe-store convert`, `moe-store inspect`; normative
  `docs/store-format-v2.md`

**`EfficientMoE/MoE-Infinity` (this repo) becomes a pure offloading engine
with zero model-arch knowledge:**

- Keeps `core/prefetch/`, `core/memory/`, `core/parallel/`, `core/model/`,
  bindings, all CUDA kernels (`extensions/`, `_v4_fp4`, `_marlin`, ...),
  Python `memory/` (tracer/predictor/policies), `engine/`, `serving/`,
  `entrypoints/`, `spec_decode/`
- Consumes moe-store via injected `TensorStore` (C++) / `TensorProvider` +
  `EngineHooks` (Python) interfaces; the two extern globals
  (`kTensorIndex`, `kArcherTensorHandle`) are deleted
- C++ shared as **source dependency** (CMake FetchContent pinned tag), never
  a prebuilt `.so` (avoids torch CXX11 ABI skew)

**Breaking change is accepted**: v1 offload dirs are not readable by v2;
migration = re-run conversion. No shims. Dependency direction:
moe-infinity → moe-store, never reverse.

**Omni-Infinity implication** (feeds back into #222): Omni-Infinity reuses
moe-store's *store/format/conversion* layer (its dense/AdaLN groups map onto
v2 group semantics); the transfer engine (prefetch/memory pools) remains
MoE-Infinity-internal and is NOT part of the shared library.

## Execution: 5 PRs (full detail in `moe-store-split.md` §4)

Each PR below lists its QA gate as tool + steps + expected result.

### PR1 — in-tree seam (no file moves, no format change)

De-globalize `kTensorIndex`/`kArcherTensorHandle` behind an in-tree
`TensorStore` interface; split `_store` pybind internals; add Python
`boundary/` protocols; behavior identical.

**QA scenarios:**

1. Build: `pip install --no-build-isolation -e .` on the sm80 dev host →
   exits 0.
2. Tests: record baseline pass set at `6da092f`, then
   `pytest tests/python/unit tests/python/integration` and `tests/cpp/unit`
   → pass set identical to baseline.
3. Seam completeness:
   `grep -rn "kArcherTensorHandle\|kTensorIndex" core/ | grep -v tensor_store`
   → empty output.
4. Smoke (GPU host):
   `python examples/readme_example.py --checkpoint deepseek-ai/DeepSeek-V2-Lite-Chat --offload_dir /tmp/pr1-smoke`
   → exit 0, coherent generated text printed.

### PR2 — moe-store repo bootstrap (native v2 format)

Scaffold repo; move `core/aio/` → `csrc/store/aio/`; add `index_v2`,
`layout_writer`, `tensor_store.h`; move Python model surface
(`models/`, `hf_config.py`, `checkpoints.py`, `constants.py`, conversion half
of `model_offload.py`); arch-aware layout planner; CLI.

**QA scenarios:**

1. Unit: `pytest tests/` in moe-store → index v2 round-trip test passes;
   layout-planner property test over every registry arch (tiny synthetic
   configs) asserts G1–G4 and byte-exact member reconstruction from single
   `[offset, offset+total_size)` reads.
2. CLI: `moe-store convert deepseek-ai/DeepSeek-V2-Lite-Chat /tmp/store-dsl`
   and a Mixtral checkpoint → exit 0;
   `moe-store inspect /tmp/store-dsl | head -20` → prints `version=2` header
   and one group per expert per layer.
3. Packaging: CPU-only CI runner builds the wheel → green.

### PR3 — moe-infinity consumes moe-store (breaking PR)

Add dependency + FetchContent pin; delete moved files (no shims);
`model_offload.py` becomes thin provider/hooks glue; engine reads v2 groups
unconditionally (contiguity probe + per-tensor fallback + dense `sort()`
workaround deleted).

**QA scenarios:**

1. Model-agnosticism: `pytest tests/python/integration/test_mock_provider.py`
   (new) → engine boots a toy 2-layer/4-expert synthetic v2 store with **no
   transformers dependency**, generates without error.
2. e2e: `python examples/readme_example.py` on DeepSeek-V2-Lite-Chat and one
   Mixtral checkpoint, offload dirs produced by `moe-store convert` → exit 0,
   coherent output, expert cache hit-rate telemetry non-empty.
3. One-read guarantee: `TensorStore`-level unit test asserts each expert
   fetch issues exactly one pread of `total_size` at `offset` (I/O counter);
   integration re-check via counter dump after e2e run.
4. Import gate:
   `grep -rn "import transformers" moe_infinity/ --include=*.py` → no hits
   in engine paths (`memory/`, `distributed/`, `engine/`); serving-layer
   tokenizer usage enumerated in the PR description.
5. Full CI: `pytest tests/python/{unit,integration,serving}` +
   `tests/cpp/unit` green.

### PR4 — DeepSeek-V4 Path-B unification onto v2

Converter handles V4 FP4 checkpoints (ExpertBundle → one group, G4);
`expert_bundle.py` safetensors reads replaced with `ReadGroup`.

**QA scenarios:**

1. `pytest tests/python/v4` against a v2 store → green.
2. V4 container harness A/B (per `moe_infinity/models/deepseek_v4/README.md`):
   Path-B on v2 store vs pre-split Path-B → logits within existing harness
   tolerance; TTFT/decode within ±5%.

### PR5 — release wiring

Publish both wheels (resolves #114 placeholder as a side effect); version
pinning `moe-infinity X.Y` ⇒ `moe-store ~=X.Y`.

**QA scenario (clean host, ≥16GB-VRAM sm80/90 GPU + SSD):** run the
executable block from `moe-store-split.md` §PR5 verbatim — fresh venv,
`pip install moe-store==<REL> moe-infinity==<REL>`, version print,
`moe-store convert` + `inspect` (expect `version=2` groups), then the
quickstart smoke snippet → exits 0 and prints `SMOKE_OUTPUT:` with non-empty
generated text (asserted in-script).

## Sequencing & risks (summary; full list in plan §5)

- PR1 → PR2 → PR3 → PR4 → PR5. PR1 lands ASAP to shrink the `models/`
  conflict window (~5 PRs/week churn); short merge freeze on `models/`
  between PR2 cut and PR3 merge.
- Top risk: V4 Path-B entanglement — isolated in PR4 behind the A/B harness;
  documented fallback if it slips (Path-B reads safetensors from inside
  moe-store temporarily, flagged explicitly).
- Out of scope: multi-node stores, PyCapsule binary ABI, prefix-cache work,
  perf work beyond the one-read guarantee.

## Alternatives Considered

1. Copy-then-diverge into Omni-Infinity — rejected in #222.
2. Generic "storage/transfer mechanism" library (memory pools + prefetch +
   H2D move out; arch knowledge stays) — the first draft of this issue;
   withdrawn: contradicts the locked plan, and the engine's prefetch/memory
   code is MoE-policy-entangled (predictor-driven) while the actual reusable
   asset is the store format + conversion + registry.
3. Monorepo — rejected in #222.

## Additional Context

- Authoritative plan: `.sisyphus/plans/moe-store-split.md` (in-repo).
- Decisions driving this: #222. Companion multimodal MoE feature: #221
  (unaffected; its input plumbing sits above the store).

## Importance

Important

## Usage Statistics (Optional)

Maintainer-driven; prerequisite (step 0) for Omni-Infinity bootstrap (#222).

**Comment by drunkcoding** [comment]
**Revision after plan review (r2, approved):** the original boundary proposal (moving `core/memory`/`core/prefetch`/transfer machinery into the shared library) is **withdrawn** — it contradicted the pre-existing locked plan `.sisyphus/plans/moe-store-split.md`. This issue is now the tracking issue that plan requires (§5): moe-store = ckpt conversion + registry + expert-id parsing + Sync* wrappers + `core/aio` store with the arch-aware v2 one-read-per-group format; MoE-Infinity keeps the entire transfer/prefetch engine. Per-PR executable QA gates added for all 5 PRs.

**Comment by drunkcoding** [comment]
**Status correction — most of this is already shipped** (this issue was drafted against a stale `dev` checkout, 32 commits behind main):
- PR1 seam: ✅ merged as #214
- PR2 repo: ✅ `EfficientMoE/moe-store` exists (tags v0.1.0, v0.2.0)
- PR3 consume: ✅ merged as #216 (model code moved out; engine serves v2 stores; `moe_infinity/models/` deleted, `boundary/` in place)
- PR4 V4 unification: ✅ `moe_store/wrappers/deepseek_v4/` present in moe-store
- PR5 release wiring: 🔶 partial (#219 pin bump, #220 CI install fix); official wheels (#114) still outstanding

**Remaining scope of this tracking issue:** (a) audit main against the locked plan's PR3/PR4 QA gates (mock-provider test, transformers-import grep gate, one-read-per-expert assertion) and close any gaps; (b) PR5 wheel publication per the §PR5 clean-host verification block.

**Comment by drunkcoding** [comment]
**QA-gate audit complete** on branch `chore/split-qa-gates` (2 commits, not yet pushed):
| Gate | Result |
|---|---|
| (a) legacy-globals grep | PASS after removing unused `kTensorIndexVersion` + regression test (`ab0f1db`) |
| (b) mock-provider model-agnosticism | PASS (`tests/python/integration/test_v2_store_engine.py`, no transformers import) |
| (c) transformers-import gate | PASS for engine paths (memory/, distributed/, engine/ = 0 hits); 5 hits remain in `runtime/model_offload.py` (lines 23, 1050, 1058, 1634, 1642) — needs triage |
| (d) one-read-per-expert | was FAIL: disk→GPU path still did per-tensor reads; fixed to single group read + Python source-contract test + C++ pread-count test (`658f35b`). C++ CTest config blocked by local CUDA/CMake env (`CUDA_CUDART_LIBRARY`); needs a GPU-host e2e run before merge |
| (e) suites | pre-existing failures on main: unit 773P/26F, offline integration 103P/5F/4E (FlashInfer + topology parity `(3,True,4)!=(3,True,3)` — pre-date this branch) |
| (f) PR5 wheels | MISSING: `pip index versions` finds neither `moe-store` nor `moe-infinity` on PyPI — #114 still open |

Remaining: publish wheels (PR5), triage `model_offload.py` transformers imports, GPU-host validation of the one-read change, and the pre-existing topology parity mismatch (may implicate moe-store planner/index).
